### PR TITLE
add script for building pythons

### PR DESCRIPTION
**Pull request by asottile-sentry**
this appears to work on:
- macos arm64
- linux aarch64

I will be following up with CI, validation

### CODE REVIEW

File: build_binary.py [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+    with tarfile.open(tgz) as tarf:
+        members = []
+        for member in tarf.getmembers():
+            _, _, member.path = member.path.partition('/')
+            members.append(member)
+        tarf.extractall(target, members=members)
+
+
+def _build(build_dir: str, prefix: str) -> int:
+    if subprocess.call(
```
Comment (joshuarli) [inline build_binary.py:380]: Could make a note to switch on `--disable-test-modules` for 3.10+?

File: build_binary.py [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+        (
+            './configure',
+            '--prefix', prefix,
+            '--without-ensurepip',
+            '--enable-optimizations',
+            '--with-lto',
```
Comment (joshuarli) [inline build_binary.py:386]: `--without-doc-strings` would be nice?

File: build_binary.py [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+def _linux_archive_name(version: Version) -> str:
+    _, libc = platform.libc_ver()
+    libc = libc.replace('.', '_')
+    return f'python-{version.s}-manylinux_{libc}_{platform.machine()}.tgz'
+
+
+BREW = '/opt/homebrew/bin/brew'
+BREW_PKG_CONFIG = '/opt/homebrew/bin/pkg-config'
+BREW_SSL = 'openssl@1.1'
+BREW_TK = 'tcl-tk'
```
[the remark this answers is not in the document]
Comment (joshuarli) [reply build_binary.py:194]: I see you have a comment below about removing tk - if possible would be nice to just not build it.

File: build_binary.py [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+        tarf.extractall(target, members=members)
+
+
+def _build(build_dir: str, prefix: str) -> int:
+    if subprocess.call(
+        (
+            './configure',
+            '--prefix', prefix,
+            '--without-ensurepip',
+            '--enable-optimizations',
```
Comment (joshuarli) [inline build_binary.py:385]: What compiler's being used? I assume gcc's in the builder image, and if so, can we pin it to a specific version (gcc-12?)

File: build_binary.py [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+
+PYTHONS = {
+    Version(3, 8, 13): Python(
+        url='https://www.python.org/ftp/python/3.8.13/Python-3.8.13.tar.xz',
+        sha256='6f309077012040aa39fe8f0c61db8c0fa1c45136763299d375c9e5756f09cf57',  # noqa: E501
+    ),
+}
+
+
+IMAGE_NAME = f'build-binary-{platform.machine()}'
```
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply build_binary.py:56]: it takes 95ns -- an lru_cache'd function takes 24ns -- it makes essentially no difference when this script is several minutes to run

File: build_binary.py [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+DOCKERFILE = f'''\
+FROM quay.io/pypa/manylinux_2_24_{platform.machine()}
+RUN : \
+    && apt-get update \
+    && DEBIAN_FROTEND=noninteractive apt-get install \
```
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply build_binary.py:61]: ah good call -- it matters sometimes (in the foreground it'll prompt for timezone setup annoyingly)

File: build_binary.py (hunk 4 above)
Comment (asottile-sentry) [reply build_binary.py:385]: for linux it's a pretty old version of gcc (6.3.0-18+deb9u1: linked to the libc version in the image) -- if we're more confident that we run on modern linuxes we can probably ditch the manylinux image and use like `ubuntu:bionic` or `ubuntu:focal` (which would change our minimum libc version)

File: build_binary.py (hunk 2 above)
Comment (asottile-sentry) [reply build_binary.py:386]: from measurements it makes only a few KB of difference -- and it breaks `help` for any builtin function

File: build_binary.py (hunk 1 above)
Comment (asottile-sentry) [reply build_binary.py:380]: ooh nice -- though I think they're still necessary for PGO (it runs a subset of the tests as part of the build)

File: build_binary.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply build_binary.py:194]: yeah I'd love to just kill tk -- afaict we don't need it

I'll remove it for now and we can add it back if we need it later

File: build_binary.py [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+    Version(3, 8, 13): Python(
+        url='https://www.python.org/ftp/python/3.8.13/Python-3.8.13.tar.xz',
+        sha256='6f309077012040aa39fe8f0c61db8c0fa1c45136763299d375c9e5756f09cf57',  # noqa: E501
+    ),
+}
+
+
+IMAGE_NAME = f'build-binary-{platform.machine()}'
+DOCKERFILE = f'''\
+FROM quay.io/pypa/manylinux_2_24_{platform.machine()}
```
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply build_binary.py:58]: agreed -- fortunately this is a one-off base image but we'll want to be bolting down the rest of our supply chain at some point

File: build_binary.py (hunk 4 above)
Comment (joshuarli) [reply build_binary.py:367]: In production, we run Debian buster (getsentry Dockerfile is based on `python:3.8.12-slim-buster`):

```
# ldd --version
ldd (Debian GLIBC 2.28-10) 2.28
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

**Comment by asottile-sentry** [comment]
> This looks fine to me. There's potentially some code in here that could be used as a shared Python library across the board.

indeed!  though as it's written right now it needs to be one dependencyless file so the linux build stuff works properly

> I don't know if it will matter but a lot of the files and directories are considered to have been created in the past: image

the timestamps are intentional!  that's the "reproducible tar" bits here: https://github.com/getsentry/prebuilt-pythons/pull/1/files#diff-da524db6fe8040689c21d15fefe6256f16cf263e57eafb4bae3116a3eaf28981R436

File: build_binary.py (hunk 4 above)
Comment (asottile-sentry) [reply build_binary.py:367]: cool, will update in a followup 👍

**Pull request merged**